service/smithyprototype/lexruntimeservice/deserialize.go

```diff
@@ -0,0 +1,1444 @@
+
+// ID is the middleware identifier
+func (p awsRestJson1_deserializeOpPutSession) ID() string {
+	return "awsRestJson1_deserializeOpPutSession"
+}
+
+func (p awsRestJson1_deserializeOpPutSession) HandleDeserialize(ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (
+	out middleware.DeserializeOutput, metadata middleware.Metadata, err error,
+) {
+	out, metadata, err = next.HandleDeserialize(ctx, in)
+	if err != nil {
+		return out, metadata, err
+	}
+	response, ok := out.RawResponse.(*smithyHTTP.Response)
+	if !ok {
+		return out, metadata, &aws.DeserializationError{
+			Err: fmt.Errorf("unknown transport type %T", out.RawResponse),
+		}
+	}
+
+	output, ok := out.Result.(*PutSessionOutput)
+	if !ok {
+		return out, metadata, &aws.DeserializationError{
+			Err: fmt.Errorf("unknown output result type %T", out.Result),
+		}
+	}
+
+	if err := awsRestJson1_deserializeRestPutSessionOutput(output, response); err != nil {
+		return out, metadata, &aws.DeserializationError{
+			Err: fmt.Errorf("failed to decode response with invalid REST bindings, %w", err),
+		}
+	}
+
+	// body of type blob
+	if err := awsRestJson1_deserializeJsonPutSessionOutput(output, response.Body); err != nil {
+		return out, metadata, &aws.DeserializationError{
+			Err: fmt.Errorf("failed to decode response body with invalid JSON, %w", err),
+		}
+	}
+
+	return out, metadata, err
+}
+
+func awsRestJson1_deserializeRestPutSessionOutput(v *PutSessionOutput, response *smithyHTTP.Response) error {
+	if v == nil {
+		return fmt.Errorf("unsupported deserialization of nil %T", v)
+	}
+
+	if v.ContentType != nil {
+		val := response.Header.Get("Content-Type")
+		v.ContentType = &val
+	}
+
+	if len(v.DialogState) > 0 {
+		val := response.Header.Get("x-amz-lex-dialog-state")
+		v.DialogState = types.DialogState(val)
+	}
+
+	if v.IntentName != nil {
+		val := response.Header.Get("x-amz-lex-intent-name")
+		v.IntentName = &val
+	}
+
+	if v.Message != nil {
+		val := response.Header.Get("x-amz-lex-message")
+		v.Message = &val
+	}
+
+	if len(v.MessageFormat) > 0 {
+		val := response.Header.Get("x-amz-lex-message-format")
+		v.MessageFormat = types.MessageFormatType(val)
+	}
+
+	if v.SessionAttributes != nil {
+		val := response.Header.Get("x-amz-lex-session-attributes")
+		b, err := base64.StdEncoding.DecodeString(val)
+		if err != nil {
+			return err
+		}
+		m := aws.JSONValue{}
+		err = json.Unmarshal(b, &m)
+		if err != nil {
+			return err
+		}
+		v.SessionAttributes = m
+	}
+
+	if v.SessionId != nil {
+		val := response.Header.Get("x-amz-lex-session-id")
+		v.SessionId = &val
+	}
+
+	if v.SlotToElicit != nil {
+		val := response.Header.Get("x-amz-lex-slot-to-elicit")
+		v.SlotToElicit = &val
+	}
+
+	if v.Slots != nil {
+		val := response.Header.Get("x-amz-lex-slots")
+		b, err := base64.StdEncoding.DecodeString(val)
+		if err != nil {
+			return err
+		}
+		m := aws.JSONValue{}
+		err = json.Unmarshal(b, &m)
+		if err != nil {
+			return err
+		}
```

Comment:
not a part of this PR but this is a good example how the SDK's current `aws.JSONValue` could be better. There is no way for a user to serialize the value into their own type, or provide their own serialize. When we end up replacing `JSONValue` we probably want a type that is lazy deserialize. This would give applications opportunity to deserialize the value into their own type.

Reply:
That can be added as a TODO but probably not here.